modules/widgets/instagram/instagram.php

```diff
@@ -0,0 +1,291 @@
+		if ( ! $instance['token_id'] && empty( $instance['title'] ) )
+			$instance['title'] = $this->defaults['title'];
+
+		if ( ! empty( $instance['title'] ) )
+			echo $args['before_title'] . esc_html( $instance['title'] ) . $args['after_title'];
+
+		if ( ! $instance['token_id'] ) {
+			echo 'This widget cannot make new connections to Instagram. You can install and use a third-party Instagram plugin instead. Please <a href="https://wordpress.com/help/contact/">contact us if you need help</a> setting this up.';
+			echo '<p><em>This notice is only shown to you.</em></p>';
+			echo '<a class="button-primary" target="_top" href="https://public-api.wordpress.com/connect/?action=request&amp;kr_nonce=25051df3f2&amp;nonce=84e239e897&amp;for=instagram-widget&amp;service=instagram&amp;blog=105567107&amp;kr_blog_nonce=217ce9172b&amp;magic=keyring&amp;instagram_widget_id=8">Authorize Instagram Access</a>';
+		} else {
+			$images = $this->get_images( $instance );
+
+			if ( ! is_array( $images ) ) {
+				echo '<p>' . __( 'There was an error retrieving images from Instagram. An attempt will be remade in a few minutes.', 'wpcomsh' ) . '</p>';
+			}
+			elseif ( ! $images ) {
+				echo '<p>' . __( 'No Instagram images were found.', 'wpcomsh' ) . '</p>';
+			}
+			else {
+				echo '<div class="' . esc_attr( 'wpcom-instagram-images wpcom-instagram-columns-' . (int) $instance['columns'] ) . '">' . "\n";
+				foreach ( $images as $image ) {
+					echo '<a href="' . esc_url( $image['link'] ) . '"><img src="' . esc_url( set_url_scheme( $image['url'] ) ) . '" width="' . esc_attr( (int) $image['width'] ) . '" height="' . esc_attr( (int) $image['height'] ) . '" alt="' . esc_attr( $image['title'] ) . '" title="' . esc_attr( $image['title'] ) . '" /></a>' . "\n";
+				}
+				echo "</div>\n";
+			}
+		}
+
+		echo $args['after_widget'];
+	}
+
+	/**
+	 * Outputs the widget configuration form for the widget administration page.
+	 * Allows the user to add new Instagram Keyring tokens and more.
+	 *
+	 * @param array $instance The widget instance (configuration options).
+	 */
+	public function form( $instance ) {
+		$instance = wp_parse_args( $instance, $this->defaults );
+
+		// No connection.
+		if ( ! $instance['token_id'] ) {
+			echo 'This widget cannot make new connections to Instagram. You can install and use a third-party Instagram plugin instead. Please <a href="https://wordpress.com/help/contact/">contact us if you need help</a> setting this up.';
+			$jetpack_blog_id = Jetpack::get_option( 'id' );
+			$response = Client::wpcom_json_api_request_as_user(
+				sprintf( '/sites/%d/external-services', $jetpack_blog_id )
+			);
+			$body = json_decode( $response['body'] );
+			$connect_URL = $body->services->instagram->connect_URL;
+			$url = add_query_arg(
+				array(
+					'siteurl' => site_url() . '/wp-admin/widgets.php',
+					'jetpack' => true,
+					'instagram_widget_id' => $this->number,
+				),
+				$connect_URL
+			);
```

Comment:
We may want to add some sort of param hash here so we can validate the params on the callback?

Reply:
Done